repository/src/main/java/org/alfresco/repo/search/impl/AbstractCategoryServiceImpl.java

```diff
@@ -546,6 +550,57 @@ public void deleteClassification(StoreRef storeRef, QName aspectName)
 
     public abstract List<Pair<NodeRef, Integer>> getTopCategories(StoreRef storeRef, QName aspectName, int count);
 
+    /**
+     * Creates search query parameters used to get top categories.
+     * Can be used as a base both wih SOLR and ES.
+     * @param storeRef Node store reference
+     * @param aspectName Aspect name. "cm:generalclassifiable" aspect should be used for usual cases.
+     *                   It is possible to use a custom aspect but it must have valid category property
+     * @param count Will be used as faceted results limit, when system has very many categories this must be reflecting that number
+     * @return SearchParameters to perform search for top categories.
+     */
+    protected SearchParameters createSearchTopCategoriesParameters(StoreRef storeRef, QName aspectName, int count) {
+        final AspectDefinition aspectDefinition = dictionaryService.getAspect(aspectName);
+        if(aspectDefinition == null)
+        {
+            throw new IllegalStateException("Unknown aspect");
+        }
+        final Map<QName, PropertyDefinition> aspectProperties = aspectDefinition.getProperties();
+        final Optional<QName> catProperty = aspectProperties.entrySet().stream()
+                //for backwards compatibility I'm leaving the part where we get custom category aspects
+                .filter(ap -> ContentModel.ASPECT_GEN_CLASSIFIABLE.isMatch(aspectName) || isValidCategoryTypeProperty(aspectName, ap))
+                .map(Map.Entry::getKey)
+                .findFirst();
+
+        return catProperty.map(cp -> {
+            final String field = "@" + cp;
+            final SearchParameters sp = new SearchParameters();
+            sp.addStore(storeRef);
+            sp.setQuery(cp + ":*");
+            //we only care about faceted results and don't need query results so we can limit them to minimum
+            sp.setMaxItems(1);
+            sp.setSkipCount(0);
+            final SearchParameters.FieldFacet ff = new SearchParameters.FieldFacet(field);
+            ff.setLimitOrNull(count < 0 ? null : count);
+            sp.addFieldFacet(ff);
+            return sp;
+        })
+                .orElseThrow(() -> new IllegalStateException("Aspect does not have category property mirroring the aspect name"));
+    }
+
+    /**
+     * Checks whether give aspect property definition is valid category property

```

Comment:
I think you forgot here about 'n' in `give` word.

Reply:
```suggestion
     * Checks whether given aspect property definition is valid category property

```